game-core/src/test/java/games/strategy/triplea/delegate/battle/MustFightBattleExecutablesTest.java

```diff
@@ -0,0 +1,378 @@
+    when(mockUnit2Attachment.getIsAaForCombatOnly()).thenReturn(true);
+
+    battle.setUnits(
+        List.of(unit2),
+        List.of(unit1),
+        List.of(),
+        List.of(),
+        defender,
+        List.of()
+    );
+    final List<IExecutable> execs = battle.getBattleExecutables(true);
+
+    assertThat(
+        "FireOffensiveAaGuns should be the first step",
+        getIndex(execs, MustFightBattle.FireOffensiveAaGuns.class), is(0));
+
+    assertThat(
+        "FireDefensiveAaGuns should be the first step",
+        getIndex(execs, MustFightBattle.FireDefensiveAaGuns.class), is(1));
+
+    assertThat(
+        "ClearAaWaitingToDieAndDamagedChangesInto is after FireOffensiveAaGuns and FireDefensiveAaGuns",
+        getIndex(execs, MustFightBattle.ClearAaWaitingToDieAndDamagedChangesInto.class), is(2));
+  }
+
+  @Test
+  @DisplayName("Verify Bombard step is added on first round")
+  void bombardStepAddedOnFirstRound() {
+    final MustFightBattle battle = newBattle(LAND);
+
+    battle.setUnits(
+        List.of(),
+        List.of(),
+        List.of(),
+        List.of(),
+        defender,
+        List.of()
+    );
+    final List<IExecutable> execs = battle.getBattleExecutables(true);
+
+    assertThat(
+        "FireNavalBombardment should be added for first round",
+        getIndex(execs, MustFightBattle.FireNavalBombardment.class), is(0));
+  }
+
+  @Test
+  @DisplayName("Verify Bombard step is not added on subsequent rounds")
+  void bombardStepNotAddedOnSubsequentRound() {
+    final MustFightBattle battle = newBattle(LAND);
+
+    battle.setUnits(
+        List.of(),
+        List.of(),
+        List.of(),
+        List.of(),
+        defender,
+        List.of()
+    );
+    final List<IExecutable> execs = battle.getBattleExecutables(false);
+
+    assertStepIsMissing(execs, MustFightBattle.FireNavalBombardment.class);
+  }
+
+  @Test
+  @DisplayName("Verify Bombard step is added but no bombardment happens if bombard units are empty")
+  void bombardStepAddedButNoBombardUnits() {
+    final MustFightBattle battle = newBattle(LAND);
+
+    battle.setUnits(
+        List.of(),
+        List.of(),
+        List.of(),
+        List.of(),
+        defender,
+        List.of()
+    );
+    final List<IExecutable> execs = battle.getBattleExecutables(true);
+
+    final int index = getIndex(execs, MustFightBattle.FireNavalBombardment.class);
+    final IExecutable step = execs.get(index);
+
+    final IDelegateBridge delegateBridge = mock(IDelegateBridge.class);
+    doAnswer(
+        invocation -> {
+          final Change change = invocation.getArgument(0);
+          mockGameData.performChange(change);
+          return null;
+        })
+        .when(delegateBridge)
+        .addChange(any());
+
+    step.execute(null, delegateBridge);
+
+    verify(mockGameData, times(1)).performChange(ChangeFactory.EMPTY_CHANGE);
+    // TODO: Somehow get the stack and check that no new execs were added to it.
```

Comment:
The `FireNavalBombardment` step adds more items to the stack.  I'd like to test that the items were either added or not added but the stack is not visible for testing nor is the `deque` inside of the stack.

I could either:
* make the stack and its deque visible to testing
* change the step code to use the stack that is currently passed in to each of the steps instead of the instance version of the stack

I prefer the later because then I can mock it but the former has less change to the code.

Reply:
Have you considered extracting the creation of the battle stack? 

If you had an object that is injected via constructor injection, you could mock it. Then you'd only need to test that you called the method and it would return an arbitrary list of battle execution steps. Then in turn you could test the usage of that class that creates the battle step stack and assert the stack has the right size and right contents in the right conditions.

It's always a bit of a code smell when you have to test via:
```
when(...).thenReturn(...)

doVoidMethodCall(..)

verify(object).someMethod(..);
```

The above is testing for side-effects, it can usually be re-arranged with a strategy object that returns a behavior or a value.

In other words, have you considered making the stack and its deque be the output of another class, a strategy class? You could then inject & mock that strategy class and then test the stack more directly.

Reply:
I have thought of that.  I'm just trying to do that in pieces.  I want to first get unittests in so that as I make the changes, I don't break something accidentally.